### PR TITLE
Docs: Draft of JSCS migration guide (fixes #5859)

### LINKED ISSUE #5859: Document transitioning from JSCS to ESLint
Once we have figured out all of the technical details to allow people to transition from JSCS to ESLint, we will need to document the process. My guess is that we'll need some documentation for transitioning both from JSCS 2.x and 3.x. The things we should cover:
1. How to convert `.jscsrc` into `.eslintrc`
2. The shareable configs that are the equivalent of JSCS presets
3. Any manual changes that need to be done in order to complete the conversion
4. An overview of how the user experience is different
5. Where to go for questions

This will clearly have to come last, just don't want to forget. :)

### PR DESCRIPTION
**Pull request by nzakas**
**What issue does this pull request address?**

#5859 

**What changes did you make? (Give an overview)**

Began the JSCS migration guide. This is just meant to be a starting point, don't get too worried if important stuff is missing. :)

**Is there anything you'd like reviewers to focus on?**

* What other terms are important to include in "Terminology"?
* What other steps will be necessary for people to convert their configs?
* What other information would be helpful for JSCS users to get acclimated to ESLint?

@markelog @mikesherov @mdevils @hzoo @zxqfox

### CODE REVIEW

File: docs/user-guide/migrating-from-jscs.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
+# Migrating from JSCS
+
+In April 2016, we [announced](http://eslint.org/blog/2016/04/welcoming-jscs-to-eslint) that the JSCS project was shutting down and the JSCS team would be joining the ESLint team. This guide is intended to help those who are using JSCS to migrate their settings and projects to use ESLint. We've tried to automate as much of the conversion as possible, but there are some manual changes that are needed.
+
+## Terminology
```
Comment (nzakas) [inline docs/user-guide/migrating-from-jscs.md]: What other terminology would be important to include here?

File: docs/user-guide/migrating-from-jscs.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+Before beginning the process of migrating to ESLint, it's helpful to understand some of the terminology that ESLint uses and how it relates to terminology that JSCS uses.
+
+* **Configuration File** - In JSCS, the configuration file is `.jscsrc`. In ESLint, the configuration file can be `.eslintrc.json`, `.eslintrc.yml`, `.eslintrc.yaml`, or `.eslintrc.js` (there is also a deprecated `.eslintrc` file format).
+* **Presets** - In JSCS, there were numerous predefined configurations shipped directly within JSCS. ESLint doesn't ship with predefined configurations, however, ESLint does support shareable configs. Shareable configs are configurations that are published on their own to npm and there are shareable configs available for almost all of the JSCS presets (see the "Converting Presets" section below). Additionally, the "preset" option in a configuration file is the equivalent of the ESLint "extends" option.
```
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: Maybe instead of "numerous" we can say 10?

File: docs/user-guide/migrating-from-jscs.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+To convert your configuration file, pass in the location of your `.jscs` file using the `--jscs` flag:
+
+```
+$ polyjuice --jscs .jscsrc > .eslintrc
+```
+
+This creates a `.eslintrc` with the equivalent rules from `.jscsrc`.
+
+If you have multiple `.jscsrc` files, you can pass them all and Polyjuice will combine them into one `.eslintrc` file:
```
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: Not sure if should mention multiple configs (is this supported already?). Not sure for what use-case that would work

File: docs/user-guide/migrating-from-jscs.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+[Polyjuice](https://github.com/brenolf/polyjuice) is a utility for converting JSCS (and JSHint) configuration files into ESLint configuration files automatically. It understands the equivalent rules from each utility and will automatically output an ESLint configuration file that is a good approximation of your existing JSCS file.
+
+To install Polyjuice:
+
+```
+$ npm install -g polyjuice
+```
+
+To convert your configuration file, pass in the location of your `.jscs` file using the `--jscs` flag:
```
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: I think we should really mention that polyjuice might not provide 1:1 mapping, so generated config probably will yield new errors so it should be expected with the eyes open as they say

File: docs/user-guide/migrating-from-jscs.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+```
+$ polyjuice --jscs .jscsrc > .eslintrc
```
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: Do you want to use deprecated extension?

File: docs/user-guide/migrating-from-jscs.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+```json
+{
+    "preset": "airbnb"
+}
+```
+
+In order to get the same functionality in ESLint, you would first need to install the `eslint-config-airbnb` shareable config package:
+
+```
+$ npm install eslint-config-airbnb --save-dev
```
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: That wouldn't work like that though

I tried it with
```sh
npm i eslint-config-airbnb eslint-plugin-import@latest eslint-plugin-jsx-a11y@latest --save-dev
```

as error output while installing `eslint-config-airbnb` suggested, but that didn't work either :/.

This kinda works -
```sh
npm install eslint-config-airbnb-base eslint-plugin-import@latest --save-dev
```

With `"extends": "airbnb"` -> "extends": "airbnb-base". Although i'm not sure if this is fare replacement.
/cc @ljharb 

Also you might want to mention if plugin is installed locally, but you execute eslint globally that would also don't work. That is a big difference with jscs, since it used locally installed plugins/configs even if installed only globally.

All this could be solved with https://github.com/eslint/eslint/issues/3458 though

File: docs/user-guide/migrating-from-jscs.md (hunk 1 above)
Comment (hzoo) [inline docs/user-guide/migrating-from-jscs.md]: Although it was removed in [jscs 3.0](http://jscs.info/changelog#version-3-0-0-https-github-comjscs-devnode-jscscomparev2-11-0-v3-0-0-2016-04-14), we could mention that there is no `esnext` option and that there are `ecmaFeatures`

Some options are the same (--fix is the same).

We can mention the error levels (0,1,2 and "off", "warn", "error") vs in jscs where you just use `true` or `false` or `null`

Comment (hzoo) [inline docs/user-guide/migrating-from-jscs.md]: error suppression in jscs http://jscs.info/overview#error-suppression vs eslint http://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments 

/* jscs: enable */
// jscs: enable
// jscs:disable requireCurlyBraces
// jscs:enable requireCurlyBraces

/* eslint-disable */
/* eslint-enable */
/* eslint-disable no-alert, no-console */
// eslint-disable-line

File: docs/user-guide/migrating-from-jscs.md (hunk 2 above)
Comment (hzoo) [inline docs/user-guide/migrating-from-jscs.md]: Can also mention that eslint does have a predefined configuration through `"extends": "eslint:recommended",`

File: docs/user-guide/migrating-from-jscs.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
+# Migrating from JSCS
+
+In April 2016, we [announced](http://eslint.org/blog/2016/04/welcoming-jscs-to-eslint) that the JSCS project was shutting down and the JSCS team would be joining the ESLint team. This guide is intended to help those who are using JSCS to migrate their settings and projects to use ESLint. We've tried to automate as much of the conversion as possible, but there are some manual changes that are needed.
+
+## Terminology
+
+Before beginning the process of migrating to ESLint, it's helpful to understand some of the terminology that ESLint uses and how it relates to terminology that JSCS uses.
+
+* **Configuration File** - In JSCS, the configuration file is `.jscsrc`. In ESLint, the configuration file can be `.eslintrc.json`, `.eslintrc.yml`, `.eslintrc.yaml`, or `.eslintrc.js` (there is also a deprecated `.eslintrc` file format).
```
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: There could more types of configs too, like `package.json`, `.jscsrc`, `.jscs.json`, `.jscs.yaml` and `.js` requires through configuration API were also supported. 

There was also `--preset=...` flag – http://jscs.info/overview#-preset-p

File: docs/user-guide/migrating-from-jscs.md (hunk 1 above)
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: > Some options are the same

Yeah, we could mention every corresponding flags

File: docs/user-guide/migrating-from-jscs.md (hunk 2 above)
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: Yeah, thought about it too, might work

File: docs/user-guide/migrating-from-jscs.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+| `airbnb`        | [`eslint-config-airbnb`](https://npmjs.com/package/eslint-config-airbnb) |
+| `crockford`        | (not available) |
+| `google`        | [`eslint-config-google`](https://npmjs.com/package/eslint-onfig-google) |
+| `grunt`        | [`eslint-config-grunt`](https://npmjs.com/package/eslint-config-grunt) |
+| `idiomatic`        | [`eslint-config-idiomatic`](https://npmjs.com/package/eslint-config-idiomatic) |
+| `jquery`        | [`eslint-config-jquery`](https://npmjs.com/package/eslint-config-jquery) |
+| `mdcs`        | (not available) |
+| `node-style-guide`        | [`eslint-config-node-style-guide`](https://npmjs.com/package/eslint-config-node-style-guide) |
+| `wikimedia`        | [`eslint-config-wikimedia`](https://npmjs.com/package/eslint-config-wikimedia) |
+| `wordpress`        | [`eslint-config-wordpress`](https://npmjs.com/package/eslint-config-wordpress) |
```
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: You explicitly want to mention, npm links not github ones?

File: docs/user-guide/migrating-from-jscs.md (hunk 5 above)
Comment (nzakas) [inline docs/user-guide/migrating-from-jscs.md]: I'm not sure what format Polyjuice outputs to. Maybe this can be eslintrc.json?

File: docs/user-guide/migrating-from-jscs.md (hunk 6 above)
Comment (nzakas) [inline docs/user-guide/migrating-from-jscs.md]: Hmm good point. Maybe recommending airbnb-base is better here. #3458 will take a while to figure out.

File: docs/user-guide/migrating-from-jscs.md (hunk 7 above)
Comment (nzakas) [inline docs/user-guide/migrating-from-jscs.md]: Ooh didn't realize that, I'll update.

File: docs/user-guide/migrating-from-jscs.md (hunk 8 above)
Comment (nzakas) [inline docs/user-guide/migrating-from-jscs.md]: I don't have the energy to hunt down all the GitHub links. If you want to do that, I'll happily merge them in.

Comment (hzoo) [inline docs/user-guide/migrating-from-jscs.md]: https://github.com/airbnb/javascript/tree/master/packages/eslint-config-airbnb
none
https://github.com/google/eslint-config-google
https://github.com/markelog/eslint-config-grunt
https://github.com/jamespamplin/eslint-config-idiomatic
https://github.com/jquery/eslint-config-jquery
n/a
https://github.com/pdehaan/eslint-config-node-style-guide
https://github.com/markelog/eslint-config-wikimedia
https://github.com/ntwb/eslint-config-wordpress

File: docs/user-guide/migrating-from-jscs.md (hunk 6 above)
Comment (ljharb) [inline docs/user-guide/migrating-from-jscs.md]: You should definitely be using airbnb-base unless you're using react - with eslint-config-airbnb you also need eslint-plugin-react and eslint-plugin-jsx-a11y.

File: docs/user-guide/migrating-from-jscs.md (hunk 1 above)
Comment (ilyavolodin) [inline docs/user-guide/migrating-from-jscs.md]: Ohh... @hzoo Is bringing up a good point. We never covered inline conversion automatically. Should it be something we have to consider for successful migration?

Comment (vitorbal) [inline docs/user-guide/migrating-from-jscs.md]: If we can read the mapping of rules from polyjuice, maybe we could provide a jstransform for this?

File: docs/user-guide/migrating-from-jscs.md (hunk 8 above)
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: @hzoo you're the man :)

File: docs/user-guide/migrating-from-jscs.md (hunk 5 above)
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: It should output json yeah, and since yaml based on (cause `.jscsrc` could be in yaml format) json compatible we should be good here, but just in case – /cc @brenolf

File: docs/user-guide/migrating-from-jscs.md (hunk 1 above)
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: Wouldn't work i think, jstransform output, as far as i can tell, is not based on tokens list but on AST i.e. it will mangle the whole file while changing the comments. It could be a eslint rule which would use polyjuice API with autofixable compatibilities. But this is seems as very unrewarding task, not sure if we could find contributors for that. Seems we are out of luck on this one :/

Comment (hzoo) [inline docs/user-guide/migrating-from-jscs.md]: You mean jscodeshift right? It wouldn't be fun but you can - POC would be something like https://astexplorer.net/#/mLIOTib1pc but applied to all nodes.

Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: https://astexplorer.net/#/U7bRRvfQXh

See my point?

Comment (hzoo) [inline docs/user-guide/migrating-from-jscs.md]: Yeah that's because https://github.com/babel/babel/pull/3561 is not in + the recast changes. the generator just prints out it's own spacing. This is the babel transform not the jscodeshift one

Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: That's doesn't really matter, lebab will not output exactly the same code

Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: Or recast for that matter, although we could use CST, hm

Comment (vitorbal) [inline docs/user-guide/migrating-from-jscs.md]: I think jscodeshift does a pretty decent job at preserving the original file. Worst case scenario, we could just warn the users that the transformation cannot be 100% trusted and that they should double-check the diff manually after the transform runs. I'm totally willing to take lead on this, perhaps we could at least experiment with an extended POC like the one @hzoo posted above!
Let me know what you guys think :)

Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: > I think jscodeshift does a pretty decent job at preserving the original file. Worst case scenario, we could just warn the users that the transformation cannot be 100% trusted and that they should double-check the diff manually after the transform runs.

Aaah, if we could say that, eslint wouldn't be bother itself with token manipulation. But with recast it might work though, since it would change only the comments.

Sounds like though, it should be a separate tool or part of the polyjuice, you still would have to use their api

Comment (vitorbal) [inline docs/user-guide/migrating-from-jscs.md]: Here is a small prototoype using jscodeshift (recast):
https://astexplorer.net/#/LjSW9EYcUP/2
Obviously it's doing it in a stupid way but just for the sake of an example.
Also, I agree that this should be a separate tool or part of polyjuice.

Comment (nzakas) [inline docs/user-guide/migrating-from-jscs.md]: @hzoo great notes, thanks, I'll make sure to update those.

For the inline comments - I think it's fair to say we can't automatically convert everything for people. If we don't have that, I think we can live without it.

File: docs/user-guide/migrating-from-jscs.md (hunk 5 above)
Comment (brenolf) [inline docs/user-guide/migrating-from-jscs.md]: Actually the input for both jscs and jshint must be json files. However the output is always a json, so we can use it as a eslintrc.json

File: docs/user-guide/migrating-from-jscs.md (hunk 3 above)
Comment (nzakas) [inline docs/user-guide/migrating-from-jscs.md]: The readme seems to indicate that it works, but if you don't think it's useful, I'm fine with removing.

File: docs/user-guide/migrating-from-jscs.md (hunk 4 above)
Comment (nzakas) [inline docs/user-guide/migrating-from-jscs.md]: Yeah, that's a good caveat to include.

File: docs/user-guide/migrating-from-jscs.md (hunk 5 above)
Comment (nzakas) [inline docs/user-guide/migrating-from-jscs.md]: Awesome, I'll update that.

File: docs/user-guide/migrating-from-jscs.md (hunk 8 above)
Comment (nzakas) [inline docs/user-guide/migrating-from-jscs.md]: For bonus points, can you send a PR to this branch with those links? :)

File: docs/user-guide/migrating-from-jscs.md (hunk 3 above)
Comment (markelog) [inline docs/user-guide/migrating-from-jscs.md]: Since we have both example uses here - with single and multiple configs, I suppose it wouldn't hurt :)

File: docs/user-guide/migrating-from-jscs.md (hunk 2 above)
Comment (nzakas) [inline docs/user-guide/migrating-from-jscs.md]: I'll make those changes.

File: docs/user-guide/migrating-from-jscs.md (hunk 8 above)
Comment (nzakas) [inline docs/user-guide/migrating-from-jscs.md]: I'm adding this now.

**Comment by nzakas** [comment]
Pushed an update. It's still not done, though I'm thinking it might be best to land something and then let people make incremental improvements rather than having me try to incorporate everyone's feedback. Thoughts?

**Comment by hzoo** [comment]
Yeah I think that's a good idea, it can be hard to follow all the comments everywhere
